### ISSUE TITLE
Numeric coercion leads to surprising results with division

### ISSUE THREAD

**Issue by jimblandy** [gpuweb/gpuweb#3326]
As far as I can tell, the following expression:
```
textureSample(t, s, vec2(1/2, 1/2))
```
actually samples for the point at `vec2(0,0)`, since the division happens in `AbstractInt`. Only when the `coords` argument is coerced to `vec2<f32>` does the possibility of `0.5` arise, which is too late.

This is one non-editorial difference between the coercion-based approach and the inference-based approach I sketched in #2651. An inference-based language would determine that, since `vec2<f32>` is the only suitable type for `textureSample`'s third argument, `vec2(1/2, 1/2)` must be `a vec2<f32>`, and thus `1/2` must be an `f32`, and thus `1` and `2` must be `f32`, and thus the division takes place in that type.

The evaluation above seems like an unfortunate outcome to me, but at this stage, unless there is great enthusiasm for fixing it, I would not expect the committee to want to reopen the discussion. I just wanted to make sure we're aware of this (quite surprising, to me) consequence of the choice we made.

**Comment by ben-clayton** [comment]
> The evaluation above seems like an unfortunate outcome to me, but at this stage, unless there is great enthusiasm for fixing it, I would not expect the committee to want to reopen the discussion. I just wanted to make sure we're aware of this (quite surprising, to me) consequence of the choice we made.

You are correct that this would result in `vec2(0,0)`. Maybe because I'm more familiar with C / C++ languages (and of course go!), this is the intuitive result to me - dividing two integer-looking literals results in an integer value. If you want a float division, make one of / both sides floating point.

Does Rust produce `0.5` in this case? I tried experimenting but [I couldn't get the compiler](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=475723b5bd6b0f95b1b4adac3410120b) to evaluate `1/2` as floating point. I know practically no Rust though.

**Comment by dneto0** [comment]
Yes, this is known. (Well, I know it...)

My concerns with #2651 still stand. See [this](https://github.com/gpuweb/gpuweb/pull/2651#discussion_r823139879): I am against a top-down approach to type inference.  I like what's in the spec: bottom-up, and monotonic progression/inference.

**Comment by jimblandy** [comment]
> Does Rust produce 0.5 in this case? I tried experimenting but I couldn't get the compiler to evaluate 1/2 as floating point. I know practically no Rust though.

This example doesn't port to Rust. Rust makes a distinction between integral and floating-point literals. If you write `1`, that'll only ever be inferred to be an integer type. If you write `1.0`, that'll only ever be inferred to be a floating-point type.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-08-30-minutes#numeric-coercion-leads-to-surprising-results-with-division-3326-">WGSL 2022-08-30 Minutes</a></summary>

* JB: Easiest way to understand is read first paragram to get example. If is surprising is matter of taste. Don’t actually have any changes just wanted to make sure it was what everyone understood. I don’t think it’s ideal as folks often write vec2(1, 2) and will be comfortable with that making float values in some contexts. Think some folks will be surprised but we’re committed to the approach we’ve taken and if we’re all OK we can close
* MM: To understand, the surprising part is ½ does not give a half. Is there any existing shading language where ½ gives .5?
* JB: Don’t know. GLSL would forbid this as you need the . somewhere
* MM: Would be an error
* JB: Yea.
* MM: Integer division doesn’t work?
* JB: It works but wouldn’t  ….
* MM: Looking for difference between compiler error and error where result is 0
* JB: Seem to remember GLSL requiring . everywhere, so don’t know what would happen. Maybe need ivec2 or something.
* **DS: So, resolution close working as intended (or won’t fix)**
* **JB: Will handle.**

</details>
